src/main/java/org/candlepin/subscriptions/db/model/EventRecord.java

```diff
@@ -0,0 +1,197 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * Red Hat trademarks are not licensed under GPLv3. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package org.candlepin.subscriptions.db.model;
+
+import org.candlepin.subscriptions.exception.ErrorCode;
+import org.candlepin.subscriptions.exception.SubscriptionsException;
+import org.candlepin.subscriptions.json.Event;
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializationFeature;
+import com.fasterxml.jackson.databind.util.StdDateFormat;
+import com.fasterxml.jackson.datatype.jdk8.Jdk8Module;
+import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
+
+import java.time.OffsetDateTime;
+import java.util.Objects;
+import java.util.UUID;
+
+import javax.persistence.AttributeConverter;
+import javax.persistence.Column;
+import javax.persistence.Convert;
+import javax.persistence.Entity;
+import javax.persistence.Id;
+import javax.persistence.Table;
+import javax.validation.Valid;
+import javax.ws.rs.core.Response;
+
+/**
+ * DB entity for an event record.
+ *
+ * An event record consists of an ID and a JSON document with the event data.
+ */
+@Entity
+@Table(name = "events")
+public class EventRecord {
+    public static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+    static {
+        // the following allows us to distinguish between
+        // - Optional.of(null) (null value, e.g. {"foo":null}), representing a field to be cleared
+        // - null (missing field, e.g. {}), representing a field that was unaffected or unknown
+        OBJECT_MAPPER.registerModule(new Jdk8Module());
+
+        OBJECT_MAPPER.registerModule(new JavaTimeModule());
+        OBJECT_MAPPER.disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);
+        OBJECT_MAPPER.disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
+        OBJECT_MAPPER.setDateFormat(new StdDateFormat().withColonInTimeZone(true));
+        OBJECT_MAPPER.setSerializationInclusion(JsonInclude.Include.NON_NULL);
+    }
+
+    /**
+     * Create a new EventRecord given Event JSON.
+     *
+     * Generates a new random UUID if the Event JSON does not have a UUID.
+     *
+     * @param json event in JSON format
+     * @return EventRecord, populated with a deserialized event record
+     * @throws JsonProcessingException if the jsonString cannot be processed
+     */
+    public static EventRecord fromJson(String json) throws JsonProcessingException {
```

Comment:
Why are these static methods?  If EventRecord had these as instance methods, Spring could create it as a Singleton, and the ObjectMapper could be injected instead of created in a `static` block (which is awkward and makes testing harder).

Reply:
@awood for simplicity of the factory method, though I guess it might be more idiomatic Spring to define an EventRecordFactory bean... Thoughts? I am happy to pivot if you like that approach (or want to suggest another).

Reply:
I agree with the sentiment that static blocks are awkward.  

Disclaimer: I haven't made it through the rest of the PR yet for context....but here's my initial thoughts:

Did you try any of the Jackson annotations for what you're trying to do? Upon seeing references to the object mapper, my first thought was to use Jackson in general...extending StdSerializer/StdSerializer if you had to.  I think @JsonCreator might work.  Also reading your comments block `@JsonIgnoreProperties` and `@JsonInclude` came to mind for dealing with nulls/missing/empty strings I think.  (admittedly I don't remember what exactly each does off the top of my head.)

https://www.baeldung.com/jackson-annotations